Exceptions

Mraoul opened the pull request:
Made some modifications to how errors are handled. Basically issue #3 is caused because unrarlib returns an error code that is not handled properly causing an infinite loop when trying to open a rar file that has its headers encrypted (the -hp flag in rar). Further I went ahead and created specific exception classes for every dll return code that exists in dll.hpp (including adding 23 and 24). Those specific classes are raised now in _check_process_result and _check_readheader_result but since all of them are based on UnrarException, you can technically still catch that. The biggest change is that ERAR_ARCHIVE_END is an exception that needs to be caught and dealt with. All non-error exceptions (e.g., ERAR_ARCHIVE_END, ERAR_MISSING_PASSWORD, etc are given non 'Error' names such as ArchiveEnd)

I've modified rarfile.py to match and also try to raise a RuntimeError in certain situation akin to the behavior of zipfile in those similar situations.

Also, if a rar file has an encrypted header, the password needs to be set before attempting to read the header, so I added a call to set the password in the init function if it was set.

And lastly there's a small/bug flaw in extract.cpp of the unrarlib whereby it will not return an ERAR_MISSING_PASSWORD if it attempts to extract a password protected file and a password is not set (I've emailed rarlabs to verify if this is a bug or I'm just missing something). As a work around, the callback function gets called to give you the opportunity to prompt the user to get a password, I just use this opportunity to mark that the file is encrypted and no password is provided so a RuntimeError can later be raised.

Mraoul commented:
I got a response from rarlabs, the bug I mentioned is apparently a known bug that has been fixed in the dll but the sources haven't been updated yet. Assuming it gets updated soon, commit 8c153563906ca388fdff21cd12facb1347b4af83 should be unnecessary

matiasb commented:
Looks good! Thanks for your detailed PR.

Respect to the RAR lib bug, AFAIUI, it sounds ok to merge your work-around since it will cover users from previous versions and it should still work ok with the fixed update, right?

Mraoul commented:
Well sorta, the reason being, my work around is only in _ReadIntoMemory, which is not used when using any of the extract functions. In those cases, it'll throw a different error, resulting in a different exception

This is the output of extract with the current library back-end using extract:
```python
Python 2.7.3 (default, Dec 18 2014, 19:10:20)
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from unrar import rarfile
>>> rf = rarfile.RarFile('test_password.rar')
>>> rf.extract('test_file.txt')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/.local/lib/python2.7/site-packages/unrar-0.3-py2.7.egg/unrar/rarfile.py", line 306, in extract
    self._extract_members([member], path, pwd)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/unrar-0.3-py2.7.egg/unrar/rarfile.py", line 339, in _extract_members
    rarinfo = self._read_header(handle)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/unrar-0.3-py2.7.egg/unrar/rarfile.py", line 154, in _read_header
    raise BadRarFile(str(e))
unrar.rarfile.BadRarFile: Bad header data.
>>>
```

And here it is using open:
```python
Python 2.7.3 (default, Dec 18 2014, 19:10:20)
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from unrar import rarfile
>>> rf = rarfile.RarFile('test_password.rar')
>>> rf.open('test_file.txt')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/.local/lib/python2.7/site-packages/unrar-0.3-py2.7.egg/unrar/rarfile.py", line 241, in open
    return data.get_bytes()
  File "/home/ubuntu/.local/lib/python2.7/site-packages/unrar-0.3-py2.7.egg/unrar/rarfile.py", line 115, in get_bytes
    raise RuntimeError('File is encrypted, password required for extraction')
RuntimeError: File is encrypted, password required for extraction
```

With my modified lib which returns the ERAR_MISSING_PASSWORD, extract looks like:
```python
Python 2.7.3 (default, Dec 18 2014, 19:10:20)
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from unrar import rarfile
>>> rf = rarfile.RarFile('test_password.rar')
>>> rf.extract('test_file.txt')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/.local/lib/python2.7/site-packages/unrar-0.3-py2.7.egg/unrar/rarfile.py", line 306, in extract
    self._extract_members([member], path, pwd)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/unrar-0.3-py2.7.egg/unrar/rarfile.py", line 341, in _extract_members
    raise RuntimeError("File is encrypted, password required")
RuntimeError: File is encrypted, password required
```
and open looks like (note! apparently I forgot to add the exception blocks to handle this assuming the lib is updated, so I just added it):

```python
Python 2.7.3 (default, Dec 18 2014, 19:10:20)
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from unrar import rarfile
>>> rf = rarfile.RarFile('test_password.rar')
>>> rf.open('test_file.txt')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/.local/lib/python2.7/site-packages/unrar-0.3-py2.7.egg/unrar/rarfile.py", line 224, in open
    raise RuntimeError("File is encrypted, password required")
RuntimeError: File is encrypted, password required
```

You could probably add a callback to the processing of the file in _extract_members akin to ReadIntoMemory but that seems like more work than it's worth. The main reason I added the work around was because if you used read/open on a password protected file the value would show up as an empty string instead of throwing an error, which would give the impression of success.

Up to you if you think that's okay or not

matiasb commented:
Got it, thanks a lot for your PR. Merging changes.
